src/core/inductive.ml

```diff
@@ -0,0 +1,257 @@
+  in aux [] !(ind_sym.sym_type)
+
+(** [principle ss pos ind_sym cons_list] returns an induction principle which
+    is created thanks to the symbol of the inductive type [ind_sym] (and its
+    position [pos]), its constructors [cons_list] and the signature [ss]. *)
+let principle : Sig_state.t -> popt -> sym -> sym list -> term =
+  fun ss pos ind_sym cons_list ->
+
+  (* STEP 0: Define some tools which will be useful *)
+  let c = get_config ss pos in
+  let p = Bindlib.new_var mkfree "p" in
+  let app  : tbox -> tbox list -> tbox = List.fold_left _Appl in
+  let fapp : sym  -> tbox list -> tbox = fun f ts -> app (_Symb f) ts in
+  let prf_of_p : tbox list -> tbox -> tbox = fun ts t ->
+    fapp c.symb_prf [_Appl (app (_Vari p) ts) t]
+  in
+
+  (* STEP 1: Create the type of the property p *)
+  let prop = _Symb c.symb_Prop in
+  let codom ts = _Impl (fapp ind_sym ts) prop in
+  let p_type = change_codom_ind_typ pos ind_sym codom in
+
+  (* STEP 2: Create each clause according to a constructor *)
+  (* [case_of cons_sym] creates a clause according to a constructor
+     [cons_sym]. *)
+  let case_of : sym -> tbox = fun cons_sym ->
+    let rec aux : tbox list -> term -> tbox = fun xs a ->
+      match Basics.get_args a with
+      | (Symb(s), ts) ->
+          if s == ind_sym then
+            prf_of_p (List.map lift ts) (app (_Symb cons_sym) (List.rev xs))
+          else
+            fatal pos "%a is not a constructor of %a"
+              pp_symbol cons_sym pp_symbol ind_sym
+      | (Prod(a,b), _) ->
+          let (x,b) = Basics.unbind_name b in
+          let b = aux ((Bindlib.box_var x)::xs) b in
+          begin
+            match Basics.get_args a with
+            | (Symb(s), ts) ->
+                let b =
+                  (* In case of a recursive argument, we create an inductive
+                     hypothesis. *)
+                  if s == ind_sym then
+                    _Impl (prf_of_p (List.map lift ts) (Bindlib.box_var x)) b
+                  else b
+                in
+                _Prod (lift a) (Bindlib.bind_var x b)
+            | _ ->
+                fatal pos "The type of %a is not supported" pp_symbol cons_sym
+          end
+      | _ -> fatal pos "The type of %a is not supported" pp_symbol cons_sym
+    in
+    let res = aux [] !(cons_sym.sym_type) in
+    if !log_enabled then
+      log_ind "The clause of the constructor [%a] is %a"
+        Print.pp_term (Symb(cons_sym))
+        Print.pp_term (Bindlib.unbox res);
+    res
+  in
+
+  (* STEP 3: Create the conclusion of the inductive principle *)
+  let codom ts =
+    let x = Bindlib.new_var mkfree "x" in
+    let t = Bindlib.bind_var x (prf_of_p ts (_Vari x)) in
+    _Prod (fapp ind_sym ts) t
+  in
+  let conclusion = change_codom_ind_typ pos ind_sym codom in
+
+  (* STEP 4: Create the induction principle *)
+
+  let t =
+    List.fold_right (fun a b -> _Impl (case_of a) b) cons_list conclusion
```

Comment:
fold_right is not tail-recursive. Use fold_left instead by removing line 279 in handle.ml (List.rev) first.

Reply:
Yes but it will break a lot of things

Reply:
Certainly not "a lot". Perhaps there will be some List.rev to add in some places but that's it.